README
DateTime::Format::DBI

This module finds a DateTime::Format::* class that is suitable for the use with
a given DBI connection (and DBD::* driver).

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

  (one of:)	DateTime::Format::MySQL
		DateTime::Format::Oracle
  		DateTime::Format::Pg
		DateTime::Format::SQLite
		... 
		(see man page for a full list of supported
		formatter/parser modules)

COPYRIGHT AND LICENCE

Copyright (C) 2003-2010 Claus Faerber

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself. 


Changes
Revision history for Perl extension DateTime::Format::DBI.

0.040 Tue Oct 19 00:00:00 2010
	- add tests with Test::Database
	- support DT::F:: modules that are singletons,
	  fixes #61614: [patch] Oracle/Sybase support

0.036 Fri Oct 15 00:00:00 2010
	- fix warning generated by DBI (reported by Ted Byers)

0.035 Sun Aug 01 00:00:00 2010
	- fixed dependencies

0.034 Mon Jul 26 00:00:00 2010
	- added experimental support for MSSQL and Sybase.
	- use Module::Build

0.033 Thu Dec 10 00:00:00 2009
	- release of 0.032_20080913
	- minor documentation fixes

0.032_20080913 00:00:00 2008
	- added experimental support for Oracle and SQLite.
	- added Test::Pod and Test::Pod::Coverage

0.032 Sat Sep 13 00:00:00 2008
	- added support for DateTime::Format::DB2
	  (patch by Chris Weyl)

0.031 Mon Nov 03 00:00:00 2003
	- include missing dependency on DBI (>= 1.2) in 
	  Makefile.PL (reported by Joshua Hoblitt).
	- database type names are now case-insensitively.

0.03  Sun Jul 20 00:00:00 2003
	- documentation now clearly states that this is only
	  a factory and describes the basic interface of the
	  returned objects.
	- requires DBI 1.21 (or higher) and uses it to determine
	  the database type through proxy/ODBC connections.

0.02  Mon Jun 16 00:00:00 2003
	- compatible with perl 5.003

0.01  Sat Jun 07 00:00:00 2003
	- original version
	- supports DBD::mysql and DateTime::Format::MySQL
	- supports DBD::Pg and DateTime::Format::Pg


Makefile.PL
use 5.003;
use ExtUtils::MakeMaker;
# $Id$
# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.
WriteMakefile(
    'NAME'		=> 'DateTime::Format::DBI',
    'VERSION_FROM'	=> 'lib/DateTime/Format/DBI.pm', # finds $VERSION
    'PREREQ_PM'		=> {
		'DateTime' => 0.10,
		'DBI'	   => 1.21,
	},
    ($] >= 5.005 ?    ## Add these new keywords supported since 5.005
      (ABSTRACT_FROM => 'lib/DateTime/Format/DBI.pm', # retrieve abstract from module
       AUTHOR     => 'Claus Faerber <CFAERBER@cpan.org>',
       LICENSE	  => 'perl'
    ) : ()),

    $ExtUtils::MakeMaker::VERSION < 6.18 ? () : (
	'SIGN'		=> 1,
    ),

    $ExtUtils::MakeMaker::VERSION < 6.45 ? () : (
	'META_MERGE'	=> { 
	    'requires'	    => { 'perl'	=> 5.003, },
#	    'recommends'    => { 'perl'	=> 5.008003, },
        },
    ),
);

sub MY::postamble {
return <<EOF
release: release-cpan

release-cpan:
	cpan-upload \$(DISTVNAME).tar\$(SUFFIX)

EOF
}


lib/DateTime/Format/DBI.pm
package DateTime::Format::DBI;

use strict;
use vars qw ($VERSION);
use warnings;

use Carp;
use DBI 1.21;

$VERSION = '0.040';
$VERSION = eval { $VERSION };

our %db_to_parser = (
  # lowercase for case-insensitivity!
  'mysql'	=> 'DateTime::Format::MySQL',
  'pg'		=> 'DateTime::Format::Pg',
  'db2'		=> 'DateTime::Format::DB2',
  'mssql'	=> 'DateTime::Format::MSSQL',  # experimental
  'oracle'	=> 'DateTime::Format::Oracle',
  'sqlite'	=> 'DateTime::Format::SQLite',
  'sybase'	=> 'DateTime::Format::Sybase',
);

sub _get_parser {
  while(@_) {
    my $dbt = lc shift;
    return $db_to_parser{$dbt}
      if exists $db_to_parser{$dbt};
  }
  return undef;
}

sub new {
  my ($name,$dbh) = @_;
  UNIVERSAL::isa($dbh,'DBI::db') || croak('Not a DBI handle.');

# NB: Using $dbh->{Driver}->{Name} this call does not work with drivers that
# connect to multiple differnt types of databases, such as DBD::Proxy,
# DBD::ODBC, DBD::JDBC,... 
#
# DBI already has code to determine the underlying database type, which is NOT
# trivial. I don't want to duplicate that here (although it's only available
# through a private API).

# my $dbtype = $dbh->{Driver}->{Name};

  my @dbtypes = eval { DBI::_dbtype_names($dbh,0) };
  my $pclass = _get_parser(@dbtypes);

  croak("No supported database driver in '@dbtypes'")
    unless defined $pclass;

  eval "use $pclass;";
  croak("Cannot load $pclass: $@") if $@;

  ## some db formatters are singletons and don't have 'new'
  ##
  my $new = UNIVERSAL::can($pclass, 'new');
  
  my $parser; if(ref $new) {
    $parser = eval { $new->($pclass); };
    croak "Cannot create object for $pclass: $@" if $@;
  } else {
    $parser = $pclass;
  }

  foreach(('format_datetime', 'parse_datetime')) {
    croak "$pclass->$_ is missing" 
      unless UNIVERSAL::can($parser, $_)
  }

  return $parser;
}

=head1 NAME

DateTime::Format::DBI - Find a parser class for a database connection.

=head1 SYNOPSIS

  use DBI;
  use DateTime;
  use DateTime::Format::DBI;

  my $db = DBI->connect('dbi:...');
  my $db_parser = DateTime::Format::DBI->new($dbh);
  my $dt = DateTime->now();

  $db->do("UPDATE table SET dt=? WHERE foo='bar'",undef,
    $db_parser->format_datetime($dt);

=head1 DESCRIPTION

This module finds a C<DateTime::Format::*> class that is suitable for the use with
a given DBI connection (and C<DBD::*> driver).

It currently supports the following format modules:
L<IBM DB2 (DB2)|DateTime::Format::DB2>,
L<Microsoft SQL (MSSQL)|DateTime::Format::MSSQL>, 
L<MySQL|DateTime::Format::MySQL>, 
L<Oracle|DateTime::Format::Oracle>,
L<PostgreSQL (Pg)|DateTime::Format::Pg>,
L<SQLite|DateTime::Format::SQLite>, and
L<Sybase|DateTime::Format::Sybase>.

B<NOTE:> This module provides a quick method to find the correct parser and
formatter class. However, this is usually not sufficient for full database
abstraction. You will also have to cater for differences in the syntax and
semantics of SQL datetime functions (and other SQL commands).

=head1 CLASS METHODS

This module provides a single factory method:

=over 4

=item * new( $dbh )

Creates a new C<DateTime::Format::*> instance, the exact class of which depends
on the database driver used for the database connection referenced by C<$dbh>. 

=back

=head1 PARSER/FORMATTER INTERFACE

C<DateTime::Format::DBI> is just a front-end class factory that will return one
of the format classes based on the nature of your C<$dbh>.

For information on the interface of the returned parser object, please see the
documentation for the class pertaining to your particular C<$dbh>.

In general, parser classes for databases will implement the following methods.
For more information on the exact behaviour of these methods, see the
documentation of the parser class.

=over 4

=item * parse_datetime( $string )

Given a string containing a date and/or time representation from the database
used, this method will return a new C<DateTime> object.

If given an improperly formatted string, this method may die.

=item * format_datetime( $dt )

Given a C<DateTime> object, this method returns a string appropriate as input
for all or the most common date and date/time types of the database used. 

=item * parse_duration( $string )

Given a string containing a duration representation from the database used,
this method will return a new C<DateTime::Duration> object.

If given an improperly formatted string, this method may die.

Not all databases and format/formatter classes support durations; please use
L<UNIVERSAL::has|UNIVERSAL/has> to check for the availability of this method.

=item * format_duration( $du )

Given a C<DateTime::Duration> object, this method returns a string appropriate
as input for the duration or interval type of the database used.

Not all databases and parser/formatter classes support durations; please use
L<UNIVERSAL::has|UNIVERSAL/has> to check for the availability of this method.

=back

Parser/formatter classes may additionally define methods like parse_I<type> or
format_I<type> (where I<type> is derived from the SQL type); please see the
documentation of the individual format class for more information.

=head1 SUPPORT

Please report bugs and other requests to the rt tracker:
L<https://rt.cpan.org/Public/Dist/Display.html?Name=DateTime-Format-DBI>.

=head1 AUTHOR

Claus FE<auml>rber <CFAERBER@cpan.org>

=head1 LICENSE

Copyright 2003-2013 Claus FE<auml>rber.  All rights reserved.  

This program is free software; you can redistribute it and/or modify it under
the same terms as Perl itself.

The full text of the license can be found in the LICENSE file included with
this module.

=head1 SEE ALSO

L<DateTime>, L<DBI>

datetime@perl.org mailing list

http://datetime.perl.org/

=cut


Build.PL
#!/usr/bin/perl

use 5.006;
use strict;
use utf8;

use Module::Build;

my $b = Module::Build->new(
	'module_name'	=> 'DateTime::Format::DBI',
	'license'	=> 'perl',
	'dist_author'	=> 'Claus Färber <CFAERBER@cpan.org>',

	'sign'		=> 1,
	'create_license' => 1,
	'create_makefile_pl' => 'traditional',

    	'requires'	=> {
		'DateTime'	=> 0.10,
		'DBI'		=> 1.21,
	},
	'recommends'	=> {
	},
	'build_requires' => {
		'Test::More' => 0, 
		'Test::NoWarnings' => 0,
	},

	auto_features => {
		    map { make_feature($_) }
			'MySQL',
			'Pg',
			'DB2',
			'MSSQL',
			'Oracle',
			'SQLite',
			'Sybase',
		},
	'no_index' => {
		'directory' => 'eg'
	},
	'resources' => {
		'homepage' => 'http://search.cpan.org/dist/',
		'repository' => 'https://perl-date-time.svn.sourceforge.net/svnroot/perl-date-time/modules/DateTime-Format-DBI/'
	},

);

$b->create_build_script;

sub make_feature {
  my $dbname = shift;
  return ( (lc $dbname).'_support' => {
	'description'	=> 'support for DateTime::Format::'.$dbname,
	'requires'	=> { 'DateTime::Format::'.$dbname => shift || 0 }
    })
}
